---
title: "Global Coral Microbiome Project: Statistical Analysis and Modelling Workflow 2024"
authors: "Colin Howe"
output: html_document
---

## Load required libraries. 
```{r, message=FALSE, warning=FALSE} 
library(BiocManager)
library(qiime2R)
library(phyloseq)
library(tidyverse)
library(ggplot2)
library(ggpubr)
library(RColorBrewer)
library(tidyverse)
library(microbiome)
library(stats4)
library(DESeq2) # differential abundance analysis
library(vegan) # anova analysis
library(RVAideMemoire)# pairwise permutation anova
library(rstatix) # dunn_test
library(ggpicrust2)

```

## Import OTU Table Taxonomy and Tree from Qiime2 and import Metadata
```{r,message=FALSE, warning=FALSE}
comp <- "tissue"
#Import from .qza file into a phyloseq object
asv_path_name <- paste0("./",comp,"/",comp,"_feature_table.qza")
asv <- qza_to_phyloseq(features = asv_path_name)

#### Import Metadata read.table
map_path_name <- paste0("./",comp,"/",comp,"_combined_metadata.txt")
metadata <- read.table(file =map_path_name ,header=T,comment.char="", row.names=1, sep="\t")

### Import Tree file from biom output tree.nwk
tree <- read_tree("./tree.nwk")

### Import taxonomy from biom output as .tsv format using read.table
taxonomy <- read.table(file = "./taxonomy.tsv", sep = "\t", header = T ,row.names = 1)

# clean the taxonomy
##code referenced from Yan Hui: email me@yanh.org github: yanhui09
tax <- taxonomy %>%
  select(Taxon) %>% 
  separate(Taxon, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), "; ")

tax.clean <- data.frame(row.names = row.names(tax),
                        Kingdom = str_replace(tax[,1], "k__",""),
                        Phylum = str_replace(tax[,2], "p__",""),
                        Class = str_replace(tax[,3], "c__",""),
                        Order = str_replace(tax[,4], "o__",""),
                        Family = str_replace(tax[,5], "f__",""),
                        Genus = str_replace(tax[,6], "g__",""),
                        Species = str_replace(tax[,7], "s__",""),
                        stringsAsFactors = FALSE)

tax.clean[is.na(tax.clean)] <- ""
tax.clean[tax.clean=="__"] <- ""

for (i in 1:nrow(tax.clean)){
  if (tax.clean[i,7] != ""){
    tax.clean$Species[i] <- paste(tax.clean$Phyla[i], tax.clean$Species[i], sep = " ")
  } else if (tax.clean[i,2] == ""){
    kingdom <- paste("Unclassified", tax.clean[i,1], sep = " ")
    tax.clean[i, 2:7] <- kingdom
  } else if (tax.clean[i,3] == ""){
    phylum <- paste("Unclassified", tax.clean[i,3], sep = " ")
    tax.clean[i, 3:7] <- phylum
  } else if (tax.clean[i,4] == ""){
    class <- paste("Unclassified", tax.clean[i,4], sep = " ")
    tax.clean[i, 4:7] <- class
  } else if (tax.clean[i,5] == ""){
    order <- paste("Unclassified", tax.clean[i,5], sep = " ")
    tax.clean[i, 5:7] <- order
  } else if (tax.clean[i,6] == ""){
    family <- paste("Unclassified", tax.clean[i,6], sep = " ")
    tax.clean[i, 6:7] <- family
  } else if (tax.clean[i,7] == ""){
    tax.clean$Species[i] <- paste("Unclassified ",tax.clean$Genus[i], sep = " ")
  }
}

for (i in 1:nrow(tax.clean)){
  if (tax.clean[i,5] == "uncultured"){
    tax.clean$Family[i] <- paste("uncultured",tax.clean$Phyla[i], tax.clean$Class[i], sep = " ")
  }
}



# create matrix format for OTU and taxonomy table
OTU <- otu_table(as.matrix(asv), taxa_are_rows = TRUE)
tax1 = tax_table(as.matrix(tax.clean))

# Set metadata
SAMPLE <- sample_data(metadata)
sample_names(SAMPLE) <- sample_names(OTU)

# Create Working phyloseq object
main_phylo <- phyloseq(OTU,tax1,SAMPLE,tree)
main_phylo

# visualize tax table
table(tax_table(main_phylo)[,"Kingdom"])


phylo <-main_phylo %>%
  phyloseq::subset_taxa(!Kingdom %in% c("Unassigned","Unclassified d__Bacteria","d__Eukaryota")) %>%
  phyloseq::subset_taxa(!Phylum %in% c("Unclassified Unassigned","Unclassified d__Archaea","Unclassified d__Bacteria","Unclassified d__Eukaryota"))

# Remove any samples with less than 1 read
phylo = prune_samples(sample_sums(main_phylo)>1, main_phylo)
phylo
table(tax_table(main_phylo)[,"Kingdom"])
```

## Calculate DISTANCE MATRICES
```{r}
## 
bray_dis = phyloseq::distance(phylo, method="bray")
uni_dis = phyloseq::distance(phylo, method="unifrac")
wuni_dis= phyloseq::distance(phylo, method="wUnifrac")

# Create data frame for downstream analysis
dis_df <- data.frame(sample_data(phylo))


```

## Ordination from Distance Matrix and PCoA Plots 
```{r}
##Calculate the PCoA on Bray-Curtis Corals
rt.pcoa = ordinate(phylo, method="PCoA", distance=bray_dis)
# Set variables to zero for subsequent changes
#Ofra
pcoa<-0
biosample = "tissue"
bray_title <-paste0(biosample,"_Neutral Community Bray-Curtis Distance")
pcoa <- plot_ordination(phylo, rt.pcoa ,color="neutrality") +
  scale_color_manual(values=c("#E41A1C","#FF7F00","#999999"))+ geom_point(size=3) + theme_bw() + ggtitle(bray_title)+theme(plot.title = element_text(hjust = 0.5)) + stat_ellipse(level = 0.95,type = "norm",aes(group=neutrality)) + theme(legend.text=element_text(size=12)) + theme(axis.text=element_text(size=12, face = "bold"), axis.title=element_text(size=14,face="bold") + theme(text = element_text(size = 20,face = "bold"))) 

##Calculate the PCoA on Unifrac Corals
rt.pcoa2 = ordinate(phylo, method="PCoA", distance=uni_dis)
# Set variables to zero for subsequent changes
#Ofra
pcoa2<-0
pcoa2 <- plot_ordination(phylo, rt.pcoa2 ,color="neutrality") +
  scale_color_manual(values=c("#E41A1C","#FF7F00","#999999"))+  geom_point(size=3) + theme_bw() + ggtitle("Neutral Microbe Communities UNifrac Distance")+theme(plot.title = element_text(hjust = 0.5)) + stat_ellipse(level = 0.95,type = "norm",aes(group=neutrality)) + theme(legend.text=element_text(size=12)) + theme(axis.text=element_text(size=12, face = "bold"), axis.title=element_text(size=14,face="bold") + theme(text = element_text(size = 20,face = "bold"))) 

##Calculate the PCoA on Unifrac Corals
rt.pcoa3 = ordinate(phylo, method="PCoA", distance=wuni_dis)
# Set variables to zero for subsequent changes
#Ofra
pcoa3<-0
pcoa3 <- plot_ordination(phylo, rt.pcoa3 ,color="neutrality") +
  scale_color_manual(values=c("#E41A1C","#FF7F00","#999999"))+  geom_point(size=3) + theme_bw() + ggtitle("Neutral Microbe Communities \n Weighted UNifrac Distance")+theme(plot.title = element_text(hjust = 0.5)) + stat_ellipse(level = 0.95,type = "norm",aes(group=neutrality)) + theme(legend.text=element_text(size=12)) + theme(axis.text=element_text(size=12, face = "bold"), axis.title=element_text(size=14,face="bold") + theme(text = element_text(size = 20,face = "bold"))) 

```

## Differential abundance Analysis
```{r}
biosample <-"mucus"
compare = list(c("below","above"),c("below","neutral"),c("neutral","above"))

for (a in compare) {
  print(a)
## Subset samples for variables
phylo.sub <- subset_samples(phylo,neutrality %in% a)
 

## Create a deseq-formatted matrix
phylo.des <- phyloseq_to_deseq2(phylo.sub, ~neutrality)

## Caluclate geometric means prior to estimate size factor
gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}
geoMeans = apply(counts(phylo.des), 1, gm_mean)

#Estimate size factor using geometric means calculation
phylo.des = estimateSizeFactors(phylo.des, geoMeans = geoMeans)
paste(a[1])
## Set the reference level * control 
phylo.des$neutrality <- relevel(phylo.des$neutrality, ref = paste(a[1]))


##Run Deseq Command
phylo.ds = DESeq(phylo.des)

# Set the alpha thresholdd
alpha = 0.05

## Evaluate Results
res <- results(phylo.ds)

##order results by the smallest p-value
resOrdered<-res[order(res$padj, res$baseMean),] 

##how many adjusted p-values < 0.05
summary(resOrdered)
l=sum(resOrdered$padj< 0.05, na.rm=TRUE) 

##get differentially expressed genes
sig_res <- resOrdered[!is.na(resOrdered$padj),]
sig_res = sig_res[sig_res$padj < 0.05,]

# This is needed to remove an outlier. One microbe was found only twice and is a likely artifact.
#sig_res = sig_res[sig_res$baseMean >= 20,]
sig_res %>% as.data.frame() -> sig_table
sig_table$asv <- row.names(sig_table)
tax_table(phylo) %>% as.data.frame()-> tax_list
tax_list$asv <- row.names(tax_list)
inner_join(sig_table,tax_list, by = "asv") -> res_tab

# Create a table contains most significant features
taxa_sig = rownames(sig_res[1:l, ])

many_col <- c("#A6CEE3", "#33A02C","#E7298A","#FDBF6F", "#B2DF8A", "#1F78B4","#E31A1C","#6A3D9A", "#FF7F00", "#FB9A99", "#CAB2D6",
              "#B15928","#1B9E77", "#D95F02", "#FFFF99", "#7570B3" ,"#66A61E", "#E6AB02","#A6761D","black", "#377EB8",
              "#4DAF4A" ,"#FF7F00","#984EA3", "#FFFF33", "#A65628","#1B9E77", "#F781BF" ,"#D53E4F","#F46D43", "#FDAE61", "#FEE08B",
              "#E6F598","#ABDDA4","#3288BD","cyan1","#66C2A5","#E69F00", "#56B4E9","#999999","#F0E442","#0072B2","#D55E00","red3",
              "yellow","skyblue","deeppink","cyan3","coral4","darkkhaki","brown1","#A6CEE3","chocolate","darkorchid2","#FF7F00",
              "#66A61E","#FDBF6F", "#FB9A99", "#CAB2D6",  "#B15928","#1B9E77","#D95F02", "#FFFF99","#E7298A", "#7570B3","#666666",
              "#E6AB02", "brown3","#A6761D", "#377EB8", "#4DAF4A" ,"#984EA3","#FFFF33","#FF7F00","black","#A6CEE3", "#33A02C","#E7298A","#FDBF6F", "#B2DF8A", "#1F78B4","#E31A1C","#6A3D9A", "#FF7F00", "#FB9A99", "#CAB2D6",
              "#B15928","#1B9E77", "#D95F02", "#FFFF99", "#7570B3" ,"#66A61E", "#E6AB02","#A6761D","black", "#377EB8",
              "#4DAF4A" ,"#FF7F00","#984EA3", "#FFFF33", "#A65628","#1B9E77", "#F781BF" ,"#D53E4F","#F46D43", "#FDAE61", "#FEE08B","darkorchid2","#FF7F00",
              "#66A61E","#FDBF6F", "#FB9A99", "#CAB2D6",  "#B15928","#1B9E77","#D95F02", "#FFFF99","#E7298A", "#7570B3","#666666",
              "#E6AB02", "brown3","#A6761D", "#377EB8", "#4DAF4A" ,"#984EA3","#FFFF33","#FF7F00","black","#A6CEE3", "#33A02C","#E7298A","#FDBF6F", "#B2DF8A", "#1F78B4","#E31A1C","#6A3D9A", "#FF7F00", "#FB9A99", "#CAB2D6",
              "#B15928","#1B9E77", "#D95F02", "#FFFF99", "#7570B3" ,"#66A61E", "#E6AB02","#A6761D","black", "#377EB8","#FF7F00","black","#A6CEE3", "#33A02C","#E7298A","#FDBF6F", "#B2DF8A", "#1F78B4","#E31A1C","#6A3D9A", "#FF7F00", "#FB9A99", "#CAB2D6",
              "#B15928","#1B9E77", "#D95F02", "#FFFF99", "#7570B3" ,"#66A61E", "#E6AB02","#A6761D","black", "#377EB8")

#Log-fold change figure
theme_set(theme_bw())

# Set results to variables for Deseq barplot 



# Set to Order
x = tapply(res_tab$log2FoldChange, res_tab$Order, function(x) max(x))
x = sort(x,decreasing = TRUE)
res_tab$Order = factor(as.character(res_tab$Order), levels=names(x))

# Set to Family
#x = tapply(res_tab$log2FoldChange, res_tab$Family, function(x) max(x))
#x = sort(x,decreasing = TRUE)
#res_tab$Family = factor(as.character(res_tab$Family), levels=names(x))

# Deseq barplot 
de_title <- paste0(biosample,"",a[1],"",a[2],"","Differential Abundance Plot")
#DE_ofra_bar <-ggplot(data=res_tab, aes(x=Order, y=log2FoldChange, fill = Order))  +
#geom_bar(stat="identity", position = "stack") + 
#  geom_hline(yintercept = 0.0, color = "black", linewidth = 1)+ scale_fill_manual(values = c(many_col))+ labs( x=" Order", y="log2FoldChange") #+ coord_flip() + ggtitle(de_title)+ theme(plot.title = element_text(hjust = 0.5), axis.text=element_text(size=6,face="bold"),axis.title=element_text(size=12))+ theme(legend.position="none")

## Create differential Expressions Dot Plot for OFRA
de_neutral_plot <- ggplot(res_tab, aes(y=Order, x=log2FoldChange, color=Order)) + 
  geom_vline(xintercept = 0.0, color = "black", linewidth = 0.8)+ scale_color_manual(values = c(many_col)) + geom_point(size=3) + theme(legend.position="none") + ggtitle(de_title)+ theme(plot.title = element_text(hjust = 0.5))
de_title_name <- paste0(biosample,"_",a[1],"_",a[2],"_","_DE_dotPlot.pdf")
ggsave(de_neutral_plot,filename = de_title_name)
}

```

## 1. PERMANOVAs Analysis
```{r}

#env material include mucus, tissue and skeleton, and outgroups without eDNA
distance_methods <-c("bray_dis","wuni_dis","uni_dis")
set.seed(129)

# This is what is called a for loop.        
for (i in distance_methods){ 
  form <- as.formula(paste(i,"neutrality", sep="~"))
  print(form)
 adonis2(form, data=dis_df)-> result
 print(result)
}
 

## Post-hoc comparison between Coral Compartments and Location
#bray_pairwise <-pairwise.perm.manova(bray_dis, dis_df$host_scientific_name,nperm=1000) 
uni_pairwise <-pairwise.perm.manova(uni_dis,dis_df$neutrality,nperm=1000)
#wuni_pairwise <-pairwise.perm.manova(wuni_dis,dis_df$host_scientific_name,nperm=1000)

```
## Alpha Diversity BoxPlots Compartments & Family
```{r, boxplots}
## Alpha Diversity Comparative Analysis and Boxplots
## Asssign metadata
sample_data(phylo) %>% as.data.frame() -> sample_meta

## Estimate Richness 
estimate_richness(phylo, measures= c("Observed","Shannon")) %>% as.data.frame() -> alpha
alpha$sample_name_backup <- sample_meta$sample_name_backup
alpha$neutrality <- sample_meta$neutrality
alpha$expedition_number <- sample_meta$expedition_number

#krust_test
krust_test <- kruskal.test(alpha$Shannon ~ neutrality,data=alpha )
krust_test

pairwise_result <- pairwise.wilcox.test(alpha$Shannon,alpha$neutrality, p.adjust.method="BH")
pairwise_result
alpha %>% filter(neutrality=="above") %>% as.data.frame () -> above_alpha
alpha %>% filter(neutrality=="below") %>% as.data.frame () -> below_alpha
alpha %>% filter(neutrality=="neutral") %>% as.data.frame () -> neutral_alpha
median(above_alpha$Observed)
median(below_alpha$Observed)
median(neutral_alpha$Observed)

## Plot Richness values
compare = list(c("below","above"),c("below","neutral"),c("neutral","above"))
alpha_boxplot <- plot_richness(phylo, x="neutrality", color = "neutrality", title = "Tissue Neutrality Alpha Diversity", measures = ("Observed"),)+ geom_boxplot() +  theme_classic() + 
  scale_color_manual(values=c("#E41A1C","#FF7F00","#999999")) + coord_cartesian(ylim=c(0,150))+
  theme(strip.background = element_blank(), plot.title = element_text(hjust = 0.5), axis.text.x.bottom = element_text(angle = -90)) + stat_compare_means(method= "wilcox.test",comparisons = compare, symnum.args=list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05), symbols = c("****", "***", "**", "*")))

```
## BAR PLOTS: Coral Compartments
```{r}
##import neutral Model results tables

tissue_table <- read.table(file ="gcmp_qiime_input/Tissue_neutral_model.tsv",header=T,comment.char="", row.names=1, sep=",")
mucus_table <- read.table(file ="gcmp_qiime_input/Mucus_neutral_model.tsv",header=T,comment.char="", row.names=1, sep=",")
skeleton_table <- read.table(file ="gcmp_qiime_input/Skeleton_neutral_model.tsv",header=T,comment.char="", row.names=1, sep=",")
combined_table <- read.table(file ="gcmp_qiime_input/combined_neutral_table.tsv",header=T,comment.char="", row.names=1, sep=",")

sig_tissue <-filter(tissue_table, padj <=0.05)
sig_skeleton <-filter(skeleton_table, padj <=0.05)
sig_mucus <-filter(mucus_table, padj <=0.05)
sig_combined <-filter(combined_table, padj <=0.05)
filter(sig_combined, Phylum == "Proteobacteria") %>%
            mutate(Order= factor(Order),
                   Order= fct_reorder(Order,p, .desc = FALSE)) -> protobacteria



many_col <- c("#A6CEE3", "#33A02C","#E7298A","#FDBF6F", "#B2DF8A", "#1F78B4","#E31A1C","#6A3D9A", "#FF7F00","#1B9E77","#B15928", "#FB9A99", "#CAB2D6", "#377EB8","#D95F02", "#7570B3","#FFFF99","#666666" ,"#66A61E", "#E6AB02","#A6761D","black", "#CAB2D6",  "#B15928","#FFFF99", "#D95F02", "#E7298A", "#7570B3","#666666" , "#E6AB02", "brown3","#A6761D", "#377EB8", "#4DAF4A" ,"#984EA3", "#FFFF33","#FF7F00","cyan3", "#4DAF4A" ,"#FF7F00","#984EA3", "#A65628","#1B9E77", "#F781BF", "#F46D43","#ABDDA4", "#FDAE61","black", "#FEE08B","#E6F598","#3288BD","deeppink","#66C2A5","cyan1","red3", "#E69F00","#56B4E9","#999999","#F0E442","#0072B2","#D55E00","#D53E4F","yellow","skyblue","coral4","#A6CEE3","white","darkkhaki","brown1","#FFFF33","chocolate","darkorchid2","#FF7F00","#1B9E77","#66A61E","#FDBF6F", "#FB9A99","#A6CEE3", "#33A02C","#E7298A","#FDBF6F", "#B2DF8A", "#1F78B4","#E31A1C","#6A3D9A", "#FF7F00","#1B9E77","#B15928", "#FB9A99", "#CAB2D6", "#377EB8","#D95F02", "#7570B3","#FFFF99","#666666" ,"#66A61E", "#E6AB02")


proto_bar <- ggplot(data = protobacteria, aes(x = model, y = Order,size=p,color = Order)) +
         geom_point() + scale_color_manual(values = c(many_col))+ theme_minimal()+labs( x=" family", y="average relative abundance (p)") + guides(color="none")+ ggtitle("Non-Neutral Proteobacteria Bargraph") + theme(plot.title = element_text(hjust = 0.5))



 
  facet_wrap(~sample_type, scales= "free_x", nrow=1) +
  labs( x=" family", y="average relative abundance (p)") + guides(fill = guide_legend(title = "Order"))+ ggtitle("Non-Neutral Proteobacteria Bargraph") + scale_fill_manual(name=NULL, values = c(many_col)) 


all_bar <- ggplot(data = sig_combined, aes(x = model, y = p, fill = Phylum)) +
         geom_bar(stat = "identity", position="stack") + 
  facet_wrap(~sample_type, scales= "free_x", nrow=1) +
  labs( x=" family", y="average relative abundance (p)") + ggtitle("Compartment Neutral Bargraph") + scale_fill_manual(name=NULL, values = c(many_col)) 

mucus_bar <- ggplot(data = sig_mucus, aes(x = model, y = p, fill = Phylum)) +
         geom_bar(stat = "identity", position="stack") + 
  labs( x=" family", y="average relative abundance (p)") + ggtitle("Mucus Neutral Bargraph") + scale_fill_manual(name=NULL, values = c(many_col)) 

tissue_bar <- ggplot(data = sig_tissue, aes(x = model, y = p, fill = Phylum)) +
         geom_bar(stat = "identity", position="stack")  +
  labs( x=" family", y="average relative abundance (p)") + ggtitle("Tissue Neutral Bargraph") +  scale_fill_manual(name=NULL, values = c(many_col)) 

skeleton_bar <- ggplot(data = sig_skeleton, aes(x = model, y = p, fill = Phylum)) +
         geom_bar(stat = "identity", position="stack")  +
  labs( x=" family", y="average relative abundance (p)") + ggtitle("Skeleton Neutral Bargraph") +  scale_fill_manual(name=NULL, values = c(many_col)) 

## For heat Map prune significant taxa found above
phylo_prune <- prune_taxa(taxa_sig, phylo)
ps.taxa.rel.sig <- prune_samples(colnames(otu_table(phylo_prune)), phylo)

# Construct Matrix from pruned samples
phylo_matrix <- as.matrix(data.frame(otu_table(phylo_prune)))

rownames(phylo_matrix) <- as.character(tax_table(phylo_prune)[, "Family"])
metadata_sub <- data.frame(sample_data(phylo_prune))


# Define the annotation color for columns and rows
annotation_col = data.frame(
    `neutrality` = as.factor(metadata_sub$neutrality), 
    check.names = TRUE
)

annotation_row = data.frame(
    Family = as.factor(tax_table(phylo_prune)[, "Family"])
)
# ann_color should be named vectors
names(many_col) = levels(annotation_row$Family)
ann_colors = list(
  `neutrality` = c(above= "cyan1", below = "blue1"),
    family = many_col) #  ,transplant = "green"

## Complexheatmap
ofra_heat <- ComplexHeatmap::pheatmap(phylo_matrix,  
                         annotation_col = annotation_col,
                         #annotation_row = annotation_row, 
                         annotation_colors = ann_colors, scale = "row",
                          color=many_col,
                            clustering_distance_cols = "manhattan", 
                         clustering_distance_rows = "manhattan",
                         clustering_method = "ward.D2",
  cluster_cols = T, cluster_rows = T, show_colnames = F, show_rownames =T,column_title =expression(paste("Differential Abundant Bacterial Heatmap for ",italic(" Orbicella franski"))), column_title_gp = gpar(fontsize = 15),fontsize = 12) # legend_breaks=-2:2


```

## ggpicrust2 analysis
```{r}

#load libraries
library(readr)
library(ggpicrust2)
library(tibble)
library(tidyverse)
library(ggprism)
library(patchwork)

## Step 5: Uses ggpicrust2 to run comparative analysis from combined psudo
#Get user input and assign to variables
comp <-"mucus"
level <-"neutral"
# Load metadata as a tibble
map_name <- paste0("./",comp,"/",comp,"_combined_metadata.txt")
metadata <- read_delim(map_name,delim = "\t", escape_double = FALSE,trim_ws = TRUE)
colnames(metadata)[1] <-"sample_name"

subset_mapping <- metadata %>% filter(neutrality!="level")

### Load KEGG pathway abundance
kegg_name <- paste0("./",comp,"_picrust2_out_pipeline","/",comp,"_KO_pred_metagenome_unstrat.tsv")
kegg_abundance <- ko2kegg_abundance(kegg_name)

filter_kegg <- kegg_abundance %>% select(-contains("level"))

# Perform pathway differential abundance analysis (DAA) using ALDEx2 method
daa_results_df <- pathway_daa(abundance = filter_kegg, metadata = subset_mapping, group = "neutrality", daa_method = "ALDEx2",reference = "neutral")

# Filter results for ALDEx2_Welch's t test method
daa_sub_method_results_df <- daa_results_df[daa_results_df$method == "ALDEx2_Wilcoxon rank test", ]

#print(paste("Annotating KO to Kegg descriptions..."))
# Annotate pathway results using KO to KEGG conversion
kegg_annotated <-pathway_annotation(pathway = "KO", daa_results_df = daa_sub_method_results_df, ko_to_kegg = TRUE)

kegg_annotated$pathway_description[is.na(kegg_annotated$pathway_description)] <- kegg_annotated$pathway_class[is.na((kegg_annotated$pathway_description))]
kegg_annotated %>% drop_na() -> kegg_annotated

## error bar
ko_errorbar <- pathway_errorbar(abundance = filter_kegg,
           daa_results_df = kegg_annotated,
           Group = subset_mapping$neutrality,
           ko_to_kegg = TRUE,
           p_values_threshold = 1e-17,
           order = "pathway_class",
           select = kegg_annotated %>%
  arrange(p_adjust) %>%
  slice(1:20) %>%
  select("feature") %>% pull(),
           p_value_bar = TRUE,
           colors = NULL,
           x_lab = "pathway_class")

# Print out Results tables
kegg_result_name <- paste0(comp,"_","_ALDEx2_glm_KO_DE_descrip.tsv")
write.table(kegg_annotated, file =kegg_result_name,sep="\t", row.names = FALSE)


### Workflow for MetaCyc Pathway 
# Load MetaCyc pathway abundance and metadata
path_name <- paste0("./",comp,"_picrust2_out_pipeline","/",comp,"_path_abun_unstrat.tsv")
print(paste("Importing ",path_name))
metacyc_table <- read.delim(path_name,header = T,check.names = FALSE)

subset_a_n_metacyc <- metacyc_table %>% select(-contains(level))
subset_mapping <- metadata %>% filter(neutrality!=level)

# Perform pathway DAA using ALDEx2 method
metacyc_daa_results_df <- pathway_daa(abundance = subset_a_n_metacyc %>% column_to_rownames("pathway"), metadata = subset_mapping, group = "neutrality", daa_method = "ALDEx2",reference = "below")

## Filter methods
metacyc_daa_results_df <- metacyc_daa_results_df[metacyc_daa_results_df$method == "ALDEx2_Wilcoxon rank test", ]

# Annotate the results
annotated_metacyc_daa_results_df <- pathway_annotation(
  pathway = "MetaCyc",
  daa_results_df = metacyc_daa_results_df,
  ko_to_kegg = FALSE)

ann_describ <- read.table(file ="metacyc_prokaryote_descrip.tsv",header=T,comment.char="",sep="\t")
colnames(ann_describ)[1] <-"feature"
left_join(annotated_metacyc_daa_results_df,ann_describ,by="feature") %>% select(-description)-> annotated_metacyc_daa_results_df
colnames(annotated_metacyc_daa_results_df)[8] <-"description"

filtered_annotation <-annotated_metacyc_daa_results_df %>% filter(!is.na(description))
#annotated_metacyc_daa_results_df$description_class[is.na(annotated_metacyc_daa_results_df$description_class)] <- #annotated_metacyc_daa_results_df$description[is.na(annotated_metacyc_daa_results_df$description_class)]



# Filter features with p < 0.05
feature_with_NA <- filtered_annotation %>%
  filter(p_adjust < 1e-40)
feature_with_p_0.05 <- drop_na(feature_with_NA) 

# Create the heatmap
heat_p <-pathway_heatmap(
  abundance = subset_a_n_metacyc %>%
    right_join(
      filtered_annotation %>% select(all_of(c("feature","description"))),
      by = c("pathway" = "feature")
    ) %>%
    filter(pathway %in% feature_with_p_0.05$feature) %>%
    select(-"pathway") %>%
    column_to_rownames("description"),
  metadata = subset_mapping,
  group = "neutrality"
)



# Print out Results tables
metacyc_result_name <- paste0(comp,"_","ALDEx2_glm_metacyc_DE_descrip.tsv")
write.table(annotated_metacyc_daa_results_df, file =metacyc_result_name,sep="\t", row.names = FALSE)

### Workflow for MetaCyc Pathway and EC
# Load MetaCyc pathway abundance and metadata
EC_path_name <- paste0("./",comp,"_picrust2_out_pipeline","/",comp,"_EC_pred_metagenome_unstrat.tsv")
print(paste("Importing",EC_path_name))
EC_table <- read.delim(EC_path_name,header = T,check.names = FALSE)

# Perform pathway DAA using LinDA method
EC_daa_results_df <- pathway_daa(abundance = EC_table %>% column_to_rownames("function"), metadata = metadata, group = "neutrality", daa_method = "ALDEx2",reference = "neutral")

## Filter methods
EC_daa_results_df <- EC_daa_results_df[EC_daa_results_df$method == "ALDEx2_glm test", ]

print(paste("Annotating EC Enzyme pathway Description"))
# Annotate the results
annotated_EC_daa_results_df <- pathway_annotation(
  pathway = "EC",
  daa_results_df = EC_daa_results_df,
  ko_to_kegg = FALSE,organism ="eco")

# Print out Results tables
EC_result_name <- paste0(comp,"_","ALDEx2_glm_EC_DE_descrip.tsv")
write.table(annotated_EC_daa_results_df, file =EC_result_name,sep="\t", row.names = FALSE)
print(paste("finished"))

```
##Ancombc analysis
```{r}
library(phyloseq)
library(tidyverse)
library(ANCOMBC)
library(DT)
library(DESeq2)
library(ComplexHeatmap)  # For advanced heatmaps
library(dplyr)   

comp <- "mucus"
#Import from .qza file into a phyloseq object
funk_path_name <- paste0("./",comp,"_picrust2_out_pipeline","/",comp,"_picrust_function_table.qza")
asv <- qza_to_phyloseq(features = funk_path_name)

#### Import Metadata read.table
map_path_name <- paste0("./",comp,"/",comp,"_combined_metadata.txt")
metadata <- read.table(file =map_path_name ,header=T,comment.char="", sep="\t")

### Import taxonomy from biom output as .tsv format using read.table
taxonomy <- read.table(file = "metacyc_prokaryote_descrip.tsv", sep = "\t", header = T,row.names = 1)
taxonomy$Family <-row.names(taxonomy)
colnames(taxonomy)[1] <-"Phylum"
# create matrix format for OTU and taxonomy table
OTU <- otu_table(as.matrix(asv), taxa_are_rows = TRUE)
tax1 = tax_table(as.matrix(taxonomy))

# Set metadata
SAMPLE <- sample_data(metadata)
sample_names(SAMPLE) <- sample_names(OTU)

# Create Working phyloseq object
funk_phylo <- phyloseq(OTU,tax1,SAMPLE)
funk_phylo
## Assign sample data to meta_data variable
meta_data <- sample_data(funk_phylo)
# factor the neutrality to set neutral first
meta_data$neutrality = factor(meta_data$neutrality, levels = c("neutral","below","above"))
# reassign factored metadata to phyloseq object
sample_data(funk_phylo) = meta_data

## Subset between two groups
subset_phylo <- subset_samples(funk_phylo, neutrality!=("neutral"))
subset_phylo

######################## ancom analysis ##############################################
# comparison between Neutral and non-neutral taxa
set.seed(111)
output = ancombc2(data = funk_phylo, tax_level = "Phylum",
                  fix_formula = "neutrality", p_adj_method = "holm", pseudo_sens = TRUE,
                  group = "neutrality", struc_zero = TRUE, neg_lb = TRUE,
                  alpha = 0.05,pairwise = TRUE)
## Pull out results
res = output$res_pair


df_fig_neutral = res %>%
    dplyr::filter(diff_neutralityabove == 1 | 
                    diff_neutralitybelow == 1) %>%
    dplyr::mutate(lfc1 = ifelse(diff_neutralityabove == 1, 
                                round(lfc_neutralityabove, 2), 0),
                  lfc2 = ifelse(diff_neutralitybelow == 1, 
                                round(lfc_neutralitybelow, 2), 0),
                  lfc3 = ifelse(diff_neutralityabove_neutralitybelow == 1, 
                                round(lfc_neutralityabove_neutralitybelow, 2), 0)) %>%
    tidyr::pivot_longer(cols = lfc1:lfc3, 
                        names_to = "group", values_to = "value") %>%
    dplyr::arrange(taxon) 


## Assign variable names 
df_fig_neutral$group = recode(df_fig_neutral$group, 
                          `lfc1` = "Above - Neutral",
                          `lfc2` = "Below - Neutral",
                          `lfc3` = "Above- Below")
df_fig_neutral$group = factor(df_fig_neutral$group, 
                          levels = c("Above - Neutral",
                                     "Below - Neutral",
                                     "Above- Below"))

## set figure range 
lo = floor(min(df_fig_neutral$value))
up = ceiling(max(df_fig_neutral$value))
mid = (lo + up)/2

## Build Heatmap
fig_neutral = df_fig_neutral %>%
  ggplot(aes(x = group, y = taxon, fill = value)) + 
  geom_tile(color = "black") +
  scale_fill_gradient2(low = "#006CD1", high = "#994F00", mid = "white", 
                       na.value = "white", midpoint = mid, limit = c(lo, up),
                       name = NULL) +
  geom_text(aes(group, taxon, label = value, color = "black"), size = 4) +
  scale_color_identity(guide = "none") +
  labs(x = NULL, y = NULL, title = "Mucus Pairwise Log fold changes in Predicted Functional Pathways betwee Neutral and Non-Neutral taxa") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5))
fig_neutral

## Subset significant functions
da_taxa <- output$res %>% filter(diff_neutralitybelow == TRUE ) %>%
  column_to_rownames("taxon") %>% select(q_neutralitybelow,lfc_neutralitybelow,W_neutralitybelow) %>% arrange(desc(lfc_neutralitybelow))# %>% as.matrix()& q_neutralitybelow > 0.05

da_heat <- output$res %>% filter(diff_neutralitybelow == TRUE & q_neutralitybelow > 1e-10) %>%
  column_to_rownames("taxon") %>% select(q_neutralitybelow,lfc_neutralitybelow,W_neutralitybelow) %>% arrange(desc(q_neutralitybelow)) %>% as.matrix()
  colnames(da_heat) <-c("adj_p","logfold","W-Statistic")

# assign function names to taxon column
#write.table(res,file="mucus_below-above_ancombc2_table.tsv", sep="\t", row.names = FALSE)
pathway_description <- read.table(file ="./metacyc_prokaryote_descrip.tsv" ,header=T,comment.char="", sep="\t")

colnames(da_taxa)[4] <- "Family"
da_taxa$Family <-row.names(da_taxa)
da_taxa_annoted <-merge(da_taxa,pathway_description, by="Family")

# first plot
plot <- ggplot(da_taxa, aes(x=lfc_neutralitybelow,y=Family),ylim=c(-1.5,1.5))+ geom_bar(stat="identity")


p <-heatmap(da_heat, scale="row")

#### Transform sample counts
ps.taxa.rel <- transform_sample_counts(subset_phylo, function(x) x/sum(x)*100)
# Subset significant  taxa
ps.taxa.rel.sig <- prune_taxa(da_taxa$taxon, ps.taxa.rel)
# Prune taxa that have a row sums that are zero or found in less than 99% of samples
ps.taxa.pse.sub <- prune_taxa(rowSums(otu_table(subset_phylo) == 0) < ncol(otu_table(subset_phylo)) * 0.99, subset_phylo)
## Prune Samples
ps.taxa.rel.sig <- prune_samples(colnames(otu_table(ps.taxa.pse.sub)), ps.taxa.rel.sig)

## create matrix
matrix <- as.matrix(data.frame(otu_table(ps.taxa.rel.sig)))

# remove NA values
matrix[!is.finite(matrix)] <-0
rownames(matrix) <- as.character(tax_table(ps.taxa.rel.sig)[, "Family"])
metadata_sub <- data.frame(sample_data(ps.taxa.rel.sig))

# Define the annotation color for columns and rows
annotation_col = data.frame(
    neutrality = as.factor(metadata_sub$neutrality), 
    check.names = FALSE)
rownames(annotation_col) = rownames(metadata_sub$neutrality)

annotation_row = data.frame(
    Description = as.factor(tax_table(ps.taxa.rel.sig)[, "Phylum"])
)
rownames(annotation_row) = rownames(matrix)

phylum_col = many_col
names(phylum_col) = levels(annotation_row$Phylum)
ann_colors = list(neutrality = c(above = "purple", below = "yellow"))

## Run heatmap
neutral_heat<- pheatmap(matrix, scale= "row", annotation_col = annotation_col, annotation_row = annotation_row, cluster_cols = FALSE, annotation_colors = ann_colors, show_colnames = FALSE)

```

## Deseq2 
```{r}
# pairwise comparison between gut and tongue
ps.taxa.sub <- subset_samples(subset_phylo, neutrality %in% c("above", "below"))
# filter sparse features, with > 90% zeros
ps.taxa.pse.sub <- prune_taxa(rowSums(otu_table(ps.taxa.sub) == 0) < ncol(otu_table(ps.taxa.sub)) * 0.9, ps.taxa.sub)
ps_ds = phyloseq_to_deseq2(ps.taxa.pse.sub, ~ neutrality)
# use alternative estimator on a condition of "every gene contains a sample with a zero"
ds <- estimateSizeFactors(ps_ds, type="poscounts")
ds = DESeq(ds, test="Wald", fitType="parametric")
alpha = 0.05 
res = results(ds, alpha=alpha)
res = res[order(res$padj, na.last=NA), ]
taxa_sig = rownames(res[1:20, ]) # select bottom 20 with lowest p.adj values
ps.taxa.rel <- transform_sample_counts(ps, function(x) x/sum(x)*100)
ps.taxa.rel.sig <- prune_taxa(taxa_sig, ps.taxa.rel)
# Only keep gut and tongue samples
ps.taxa.rel.sig <- prune_samples(colnames(otu_table(ps.taxa.pse.sub)), ps.taxa.rel.sig)


```
## Functional output into qiime
```{r}
## Goal is to set up functional taxonomy file that can be imported with metagenome_unstrat files into phyloseq
#load libraries
library(readr)
library(ggpicrust2)
library(tibble)
library(tidyverse)
library(ggprism)
library(patchwork)

comp = "mucus"
# Load metadata as a tibble
map_name <- paste0("./",comp,"/",comp,"_combined_metadata.txt")
metadata <- read_delim(map_name,delim = "\t", escape_double = FALSE,trim_ws = TRUE)
colnames(metadata)[1] <-"sample_name"

# Load KEGG pathway abundance
#kegg_name <- paste0("./",comp,"_picrust2_out_pipeline","/",comp,"_EC_pred_metagenome_unstrat.tsv")
#kegg_abundance <- ko2kegg_abundance(kegg_name)

# Perform pathway differential abundance analysis (DAA) using ALDEx2 method
#daa_results_df <- pathway_daa(abundance = kegg_abundance, metadata = metadata, group = "neutrality", daa_method = "ALDEx2",reference = "neutral")

# Filter results for ALDEx2_Welch's t test method
#daa_sub_method_results_df <- daa_results_df[daa_results_df$method == "ALDEx2_glm test", ]

# Annotate pathway results using KO to KEGG conversion
#pathway_annotation(pathway = "KO", daa_results_df = daa_sub_method_results_df, ko_to_kegg = TRUE, organism ="eco") %>% #select(c("feature","pathway_name","pathway_map","pathway_class","pathway_description")) -> kegg_annotation_tax


# Print out Results tables
#kegg_result_name <- paste0(comp,"_","_ALDEx2_glm_KO_DE_descrip.tsv")
#write.table(kegg_annotation_tax, file =kegg_result_name,sep="\t", row.names = FALSE)

# Workflow for MetaCyc Pathway and EC
# Load MetaCyc pathway abundance and metadata
path_name <- paste0("./",comp,"_picrust2_out_pipeline","/",comp,"_path_abun_unstrat.tsv")
metacyc_table <- read.table(path_name,header = T,check.names = FALSE)

# Perform pathway DAA using ALDEx2 method
metacyc_daa_results_df <- pathway_daa(abundance = metacyc_table %>% column_to_rownames("pathway"), metadata = metadata, group = "neutrality", daa_method = "ALDEx2",reference = "neutral")

## Filter methods
metacyc_daa_results_df <- metacyc_daa_results_df[metacyc_daa_results_df$method == "ALDEx2_glm test", ]

# Annotate the results
annotated_metacyc_daa_results_df <- pathway_annotation(
  pathway = "MetaCyc",
  daa_results_df = metacyc_daa_results_df,
  ko_to_kegg = FALSE,organism ="eco")

# Print out Results tables
metacyc_result_name <- paste0(comp,"_","ALDEx2_glm_metacyc_DE_descrip.tsv")
write.table(annotated_metacyc_daa_results_df, file =metacyc_result_name,sep="\t", row.names = FALSE)

# Workflow for MetaCyc Pathway and EC
# Load MetaCyc pathway abundance and metadata
EC_path_name <- paste0("./",comp,"_picrust2_out_pipeline","/",comp,"_EC_pred_metagenome_unstrat.tsv")
EC_table <- read.delim(EC_path_name,header = T,check.names = FALSE)

# Perform pathway DAA using LinDA method
EC_daa_results_df <- pathway_daa(abundance = EC_table %>% column_to_rownames("#NAME"), metadata = metadata, group = "neutrality", daa_method = "ALDEx2",reference = "neutral")

## Filter methods
EC_daa_results_df <- EC_daa_results_df[EC_daa_results_df$method == "ALDEx2_glm test", ]

# Annotate the results
annotated_EC_daa_results_df <- pathway_annotation(
  pathway = "EC",
  daa_results_df = EC_daa_results_df,
  ko_to_kegg = FALSE,organism ="eco")

# Print out Results tables
EC_result_name <- paste0(comp,"_","ALDEx2_glm_EC_DE_descrip.tsv")
write.table(annotated_EC_daa_results_df, file =EC_result_name,sep="\t", row.names = FALSE)

```


## Parking lot
```{r}
## Subset glom OTU table
phyloseq::otu_table(glom)%>%
        as.data.frame()%>%
        rownames_to_column("id") -> glom_otu_table

## Subset glom tax table
phyloseq::tax_table(glom)%>%
        as.data.frame()%>%
        rownames_to_column("id") -> glom_otu_tax

## Subset glom tax table
phyloseq::sample_data(glom)%>%
        as.data.frame() -> glom_metadata

# Connnect taxonomy information to neutral  model results
inner_join(neutral_results_padj,glom_tax, by = "id") %>% as.data.frame() -> nonneutral_tax


#subset otu table based on neutral results
neutral_table <-subset(glom_otu_table, id  %in%  neutral_fam_df$id)

#subset taxonomy table based on neutral results
neutral_tax <- subset(glom_otu_tax, id  %in%  neutral_fam_df$id)# & Family %in% neutral_fam_df$Family)

#subset Metadata based on neutral results
neutral_sample_name <- colnames(neutral_table)
neutral_meta <-subset(glom_metadata,sample_name_backup %in% neutral_sample_name)
neutral_metadata <- sample_data(neutral_meta)

## This approch helps transpose the data without adding numbers as colummn id
test_table <- setNames(data.frame(t(glom_otu_table[,-1])),glom_otu_table[,1])

# create matrix format for OTU and taxonomy table
OTU2 <- otu_table(data.matrix(test_table, rownames.force = NA), taxa_are_rows = TRUE)
tax2 = tax_table(as.matrix(neutral_tax))

# Create Working phyloseq object
neutral_phylo <- phyloseq(OTU2,tax2,neutral_metadata)

neutral_phylo

############################# Glom Psudo Tables ##############################################################

########### Redo for neutral without filtering for  significant padj ########
phyloseq::tax_table(glom) %>%
  as.data.frame() %>% 
  rownames_to_column("id") -> Tissue_glom_tax

## Inner_join will connect  ASV id with taxonomy using id column
inner_join(neutral_results_padj,Tissue_glom_tax, by = "id") -> non_neutral_tax

## Subset ASV from each model and signifcance 
sig_neutral_tax = select(filter(non_neutral_tax, model == "neutral"),id,p_abundance,freq,padj,Phylum,Class,Order,Family,model)

## Subset glom OTU table
phyloseq::otu_table(glom)%>%
        as.data.frame()%>%
        rownames_to_column("id") -> glom_otu_table

##subset otu table based on neutral results and rename column names
neutral_table <-subset(glom_otu_table, id  %in%  sig_neutral_tax$id)
n <-length(neutral_table)
colnames(neutral_table)[2:n] <- paste(colnames(neutral_table)[2:n], "neutral", sep = "_")

## Subset metadata file
data.frame(sample_data(glom)) %>% select(sample_name_backup, expedition_number,host_scientific_name) %>% as.data.frame() -> glom_meta

## Create new columns for the metadata data
glom_meta %>% mutate(psuedo_sample_name = paste(colnames(neutral_table)[2:n]),
                         compartment_neutraility = paste("Tissue", "neutral", sep = '_'),
                         compartment = paste ("Tissue"),
                         neutraility = paste("neutral")) %>% 
                         as.data.frame() -> glom_metadata

psuedo_table_name <- paste0("Tissue","_","neutral_table.csv")
(write.csv(neutral_table, file =psuedo_table_name, row.names = FALSE))

psuedo_map_name <- paste0("Tissue","_","neutral_metadata.csv")
(write.csv(glom_metadata, file =psuedo_map_name, row.names = FALSE))

psuedo_tax_name <- paste0("Tissue","_","neutral_taxonomy.csv")
(write.csv(subset_tax, file =psuedo_tax_name, row.names = FALSE))
# Subset Taxonomy and join with neutral model results
neutral = c("above","below")
for (i in neutral){
phyloseq::tax_table(glom) %>%
  as.data.frame() %>% 
  rownames_to_column("id") -> glom_tax
## Inner_join will connect  ASV id with taxonomy using id column
inner_join(neutral_results_padj,glom_tax, by = "id") -> non_neutral_tax

## Subset ASV from each model and signifcance 
sig_neutral_tax = select(filter(non_neutral_tax, model == i & padj <=0.05),id,p_abundance,freq,padj,Phylum,Class,Order,Family,model)

## Subset glom OTU table
phyloseq::otu_table(glom)%>%
        as.data.frame()%>%
        rownames_to_column("id") -> glom_otu_table

#subset otu table based on neutral results and rename column names
neutral_table <-subset(glom_otu_table, id  %in%  sig_neutral_tax$id)
l = length(neutral_table)
colnames(neutral_table)[2:l] <- paste(colnames(neutral_table)[2:l], i, sep = "_")

## Subset metadata file
data.frame(sample_data(glom)) %>% select(sample_name_backup, expedition_number,host_scientific_name) %>% as.data.frame() -> glom_meta
## Create new columns for the metadata data
glom_meta %>% mutate(psuedo_sample_name = paste(colnames(neutral_table)[2:299]),
                         compartment_neutraility = paste("mucus", i, sep = '_'),
                         compartment = paste ("mucus"),
                         neutraility = paste(i)) %>% 
                         as.data.frame() -> glom_metadata

## Subset significant results with taxonomy to retain only significant taxonomy
subset_tax <- subset(glom_tax, id %in% sig_neutral_tax$id) %>% 
  select(-c("Genus","Species"))

psuedo_table_name <- paste0("mucus","_",i,"_table.tsv")
(write.table(neutral_table, file =psuedo_table_name, row.names = FALSE))

psuedo_map_name <- paste0("mucus","_",i,"_metadata.tsv")
(write.table(glom_metadata, file =psuedo_map_name, row.names = FALSE))

psuedo_tax_name <- paste0("mucus","_",i,"_taxonomy.tsv")
(write.table(subset_tax, file =psuedo_tax_name, row.names = FALSE))

}

########### Redo for neutral without filtering for  significant padj ########
phyloseq::tax_table(glom) %>%
  as.data.frame() %>% 
  rownames_to_column("id") -> glom_tax

## Inner_join will connect  ASV id with taxonomy using id column
inner_join(neutral_results_padj,glom_tax, by = "id") -> non_neutral_tax

## Subset ASV from each model and signifcance 
sig_neutral_tax = select(filter(non_neutral_tax, model == "neutral"),id,p_abundance,freq,padj,Phylum,Class,Order,Family,model)

## Subset glom OTU table
phyloseq::otu_table(glom)%>%
        as.data.frame()%>%
        rownames_to_column("id") -> glom_otu_table

##subset otu table based on neutral results and rename column names
neutral_table <-subset(glom_otu_table, id  %in%  sig_neutral_tax$id)
colnames(neutral_table)[2:299] <- paste(colnames(neutral_table)[2:299], "neutral", sep = "_")

## Subset metadata file
data.frame(sample_data(glom)) %>% select(sample_name_backup, expedition_number,host_scientific_name) %>% as.data.frame() -> glom_meta

## Create new columns for the metadata data
glom_meta %>% mutate(psuedo_sample_name = paste(colnames(neutral_table)[2:299]),
                         compartment_neutraility = paste("mucus", "neutral", sep = '_'),
                         compartment = paste ("mucus"),
                         neutraility = paste("neutral")) %>% 
                         as.data.frame() -> glom_metadata

psuedo_table_name <- paste0("mucus","_","neutral_table.tsv")
(write.table(neutral_table, file =psuedo_table_name, row.names = FALSE))

psuedo_map_name <- paste0("mucus","_","neutral_metadata.tsv")
(write.table(glom_metadata, file =psuedo_map_name, row.names = FALSE))

psuedo_tax_name <- paste0("mucus","_","neutral_taxonomy.tsv")
(write.table(subset_tax, file =psuedo_tax_name, row.names = FALSE))

################# ggPicrust2 #############################################
#load libraries
library(readr)
library(ggpicrust2)
library(tibble)
library(tidyverse)
library(ggprism)
library(patchwork)

comp = "mucus"
# Load metadata as a tibble
map_name <- paste0("./",comp,"/",comp,"_combined_metadata.txt")
metadata <- read_delim(map_name,delim = "\t", escape_double = FALSE,trim_ws = TRUE)
colnames(metadata)[1] <-"sample_name"

kegg_abundance %>% select(matches("above|neutral")) -> select_table
metadata %>% filter(neutrality=="above" | neutrality == "neutral") -> filter_metadata

# Load KEGG pathway abundance
kegg_name <- paste0("./",comp,"_picrust2_out_pipeline","/",comp,"_KO_pred_metagenome_unstrat.tsv")
kegg_abundance <- ko2kegg_abundance(kegg_name)

# Perform pathway differential abundance analysis (DAA) using ALDEx2 method
daa_results_df <- pathway_daa(abundance = select_table, metadata = filter_metadata, group = "neutrality", daa_method = "ALDEx2",reference = "neutral")

# Filter results for ALDEx2_Welch's t test method
daa_sub_method_results_df <- daa_results_df[daa_results_df$method == "ALDEx2_Wilcoxon rank test", ]

# Annotate pathway results using KO to KEGG conversion
daa_annotated_sub_method_results_df <- pathway_annotation(pathway = "KO", daa_results_df = daa_sub_method_results_df, ko_to_kegg = TRUE)

# Print out Results tables
kegg_result_name <- paste0(comp,"_","_ALDEx2_Wilcoxon_AN_KO_DE_descrip.tsv")
write.table(daa_annotated_sub_method_results_df, file =kegg_result_name,sep="\t", row.names = FALSE)

# Generate pathway error bar plot
# Please change Group to metadata$your_group_column if you are not using example dataset
#p <- pathway_errorbar(abundance = select_table, daa_results_df = daa_annotated_sub_method_results_df, Group = filter_metadata$neutrality, #p_values_threshold = 0.05, order = "pathway_class", select = NULL, ko_to_kegg = TRUE, p_value_bar = TRUE, colors = NULL, x_lab = "pathway_name")

# Workflow for MetaCyc Pathway and EC
# Load MetaCyc pathway abundance and metadata
path_name <- paste0("./",comp,"_picrust2_out_pipeline","/",comp,"_path_abun_unstrat.tsv")
metacyc_table <- read.table(path_name,header = T,check.names = FALSE)
## Subset between two groups
metacyc_table %>% select(matches("pathway|above|neutral")) %>% as.data.frame()-> select_metacyc_abundance

# Perform pathway DAA using LinDA method
metacyc_daa_results_df <- pathway_daa(abundance = select_metacyc_abundance %>% column_to_rownames("pathway"), metadata = filter_metadata, group = "neutrality", daa_method = "ALDEx2")


# Create the heatmap
p_heat <-pathway_heatmap(abundance = select_metacyc_abundance %>% filter(pathway %in% feature_with_p_0.05$feature) %>% column_to_rownames("pathway"), metadata = metadata, group = "neutrality") 

# Generate pathway PCA plot
pathway_pca(abundance = select_metacyc_abundance %>% column_to_rownames("pathway"), metadata = filter_metadata, group = "neutrality") -> path_pca_graph


```

```